irohad/main/impl/on_demand_ordering_init.cpp

```diff
@@ -235,31 +235,56 @@ namespace iroha {
                 return hashes;
               }),
           sync_event_notifier.get_observable()
+              .tap([this](const synchronizer::SynchronizationEvent &event) {
+                last_received_round_ = event.round;
+              })
               .lift<iroha::synchronizer::SynchronizationEvent>(
                   iroha::makeDelay<iroha::synchronizer::SynchronizationEvent>(
                       delay_func, rxcpp::identity_current_thread()))
-              .map([log = log_](const auto &event) {
+              .flat_map([this](const auto &event) mutable
```

Comment:
```c++
              .filter([this](const auto &event) {
                assert(last_received_round_);
                if (not last_received_round_) {
                  log_->error("Cannot continue without last received round");
                  return false;
                }
                if (event.round < *last_received_round_) {
                  log_->debug("Dropping {}, since {} is already processed",
                              event.round,
                              *last_received_round_);
                  return false;
                }
                return true;
              })
              .map([this](const auto &event) {
                consensus::Round current_round;
                switch (event.sync_outcome) {
                  case iroha::synchronizer::SynchronizationOutcomeType::kCommit:
                    log_->debug("Sync event on {}: commit.",
                                *last_received_round_);
                    current_round =
                        ordering::nextCommitRound(*last_received_round_);
                    break;
                  case iroha::synchronizer::SynchronizationOutcomeType::kReject:
                    log_->debug("Sync event on {}: reject.",
                                *last_received_round_);
                    current_round =
                        ordering::nextRejectRound(*last_received_round_);
                    break;
                  case iroha::synchronizer::SynchronizationOutcomeType::
                      kNothing:
                    log_->debug("Sync event on {}: nothing.",
                                *last_received_round_);
                    current_round =
                        ordering::nextRejectRound(*last_received_round_);
                    break;
                  default:
                    log_->error("unknown SynchronizationOutcomeType");
                    assert(false);
                }
                return ordering::OnDemandOrderingGate::RoundSwitch{
                    std::move(current_round), event.ledger_state};
              }),
```